src/librustc_lint/types.rs

```diff
@@ -193,9 +220,36 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TypeLimits {
                                     }
                                 }
                             }
-                            cx.span_lint(OVERFLOWING_LITERALS,
-                                         e.span,
-                                         &format!("literal out of range for {:?}", t));
+                            if let Some(repr_str) = get_bin_hex_repr(cx, lit) {
+                                let bits = uint_ty_bits(t, cx.sess().target.usize_ty);
+                                let actually = (lit_val << (128 - bits)) >> (128 - bits);
+                                let mut err = cx.struct_span_lint(
+                                    OVERFLOWING_LITERALS,
+                                    e.span,
+                                    &format!("literal out of range for {:?}", t),
+                                );
+                                err.note(&format!(
+                                    "the literal `{}` (decimal `{}`) does not fit into \
+                                     an `{:?}` and will become `{}{:?}`.",
+                                    repr_str, lit_val, t, actually, t
+                                ));
+                                let sugg_ty = get_type_suggestion(
+                                    &cx.tables.node_id_to_type(e.hir_id).sty,
+                                    lit_val,
+                                    false,
+                                );
+                                if !sugg_ty.is_empty() {
+                                    err.help(&sugg_ty);
```

Comment:
It'd be nice if this were a `span_suggestion` instead if it is a numeric literal that sets the type, otherwise provide the help (ideally I would check the source of the type constraint and point at it, but it is not necessary to go down that rabbit hole in this PR):

```
warning: literal out of range for i8
  --> $DIR/type-overflow.rs:34:17
   |
34 |     let fail = -0b1111_1111i8;
   |                 ^^^^^^^^^^^^^ help: consider using `i16` instead: `-0b1111_1111i16`
   |
   = note: the literal `0b1111_1111i8` (decimal `255`) does not fit into an `i8` and will become `-1i8`
```

Reply:
This looks better. I will add this, this evening.